project-koku/koku
# [COST-3890/3892] Data Extraction and Kafka Messaging for SUBS

## Jira Ticket

[COST-3890](https://issues.redhat.com/browse/COST-3890)
[COST-3892](https://issues.redhat.com/browse/COST-3892)

Also addresses and ask from SWATCH to send product ids appropriately based on tags
[COST-4062](https://issues.redhat.com/browse/COST-4062)

## Description

This change will add identification of RHEL metering records and processing of those records to S3, followed by another task to process S3 reports to a kafka message on a different worker.

## Testing

1. Checkout Branch
2. Restart Koku and include a subs worker `make docker-up-min-with-subs`
3. Create and Load test customer data, the AWS source yaml was updated to contain a specific tag on an EC2 generator that causes the record to be identified as RHEL metered. This tag is case insensitive. REQUIRED TAG FOR IDENTIFICATION: `resourceTags/user:com_redhat_rhel: true`
4. in the logs you should see a message that says `beginning subs rhel extraction`, this indicates it is attempting to process the AWS report and identify relevant records.
5. There will be a following log of `1 file(s) uploaded to s3 for rhel metering` if relevant records were found.
6. Once the process to kafka task has run, you will see a log of `iterating over records and sending kafka messages` followed by a count of the number of messages sent `sent 1438 kafka messages for subs`.


## Notes
LOCAL NOTES:
- Locally the bucket uploaded to is the regular minio koku-bucket. the path will be something like this with the org as a top level path in the bucket: 
```
koku-bucket/org1234567/AWS/source=fee1eb86-d751-49c4-a8ff-be1a5d2f48ea/date=2023-07-17/subs_9281881a-2963-48f4-a27b-6f5a5397bfe3_0.csv
```
- Unless you spin up kafka, you will see an error related to connecting to kafka but it will not prevent processing
